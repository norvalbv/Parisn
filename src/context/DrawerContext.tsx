import React, { createContext, ReactElement, useMemo, useState } from 'react';

export type DrawerContextProps = {
  openedId: string;
  closeDrawer: () => void;
  openDrawer: (arg: string) => void;
};

export const DrawerContext = createContext<null | DrawerContextProps>(null);

type DrawerProviderProps = {
  children: JSX.Element | JSX.Element[];
};

export const DrawerProvider = ({ children }: DrawerProviderProps): ReactElement => {
  const [openedId, setOpenedId] = useState('');

  const openDrawer = (id: string): void => {
    setOpenedId(id);
  };

  const closeDrawer = (): void => {
    setOpenedId('');
  };

  const memoisedValue = useMemo(() => {
    return {
      openedId,
      openDrawer,
      closeDrawer,
    };
  }, [openedId]);

  return <DrawerContext.Provider value={memoisedValue}>{children}</DrawerContext.Provider>;
};
