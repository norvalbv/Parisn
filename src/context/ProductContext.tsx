import React, { createContext, useState, useMemo, useEffect, ReactElement } from 'react';
import { ProductContextData, ProductInfoValues } from '../types';

type ProductContextProviderProps = {
  children?: JSX.Element;
};

const ProductContext = createContext<ProductContextData | null>(null);

export const ProductContextProvider = ({ children }: ProductContextProviderProps): ReactElement => {
  const [productInfo, setProductInfo] = useState<ProductInfoValues | null>({
    product: null,
    price: null,
    selectedSize: null,
  });

  useEffect(() => {
    const retreviedProductInfo = localStorage.getItem('savedProductInfo') || 'null';
    const parsedData = JSON.parse(retreviedProductInfo) as ProductInfoValues;
    // Checks whether all data is truthy or not.
    const truthyDataParsed = parsedData !== null && Object.values(parsedData).every((item) => item);

    if (truthyDataParsed) {
      return setProductInfo(parsedData);
    }
  }, []);

  useEffect(() => {
    localStorage.setItem('savedProductInfo', JSON.stringify(productInfo));
  }, [productInfo]);

  const memoisedValue = useMemo(
    () => ({
      productInfo,
      setProductInfo,
    }),
    [productInfo]
  );

  return <ProductContext.Provider value={memoisedValue}>{children}</ProductContext.Provider>;
};

export default ProductContext;
