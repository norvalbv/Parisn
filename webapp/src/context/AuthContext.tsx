import { CognitoUserPool } from 'amazon-cognito-identity-js';

const poolData = {
  UserPoolId: 'eu-west-2_cuz8GIlYg',
  ClientId: '3gh7ll3fdvsq6nh642g06emka6',
};

export default new CognitoUserPool(poolData);
