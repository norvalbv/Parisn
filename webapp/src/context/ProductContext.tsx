import React, { createContext, useState, useMemo, useEffect } from 'react';
import { ProductContextData, ProductInfoValues } from '../types';

type ProductContextProviderProps = {
  children?: JSX.Element;
};

const ProductContext = createContext<ProductContextData | null>(null);

export const ProductContextProvider = ({ children }: ProductContextProviderProps) => {
  const [productInfo, setProductInfo] = useState<ProductInfoValues | null>({
    product: null,
    price: null,
    selectedSize: null,
  });

  useEffect(() => {
    const retreviedProductInfo = localStorage.getItem('savedProductInfo');
    const parsedData = JSON.parse(retreviedProductInfo || 'null');
    // Checks whether all data is truthy or not.
    const truthyDataParsed = parsedData !== null && Object.values(parsedData).every((item) => item);

    if (truthyDataParsed) {
      return setProductInfo(parsedData);
    }
  }, []);

  useEffect(() => {
    localStorage.setItem('savedProductInfo', JSON.stringify(productInfo));
  }, [productInfo]);

  const memoisedValue = useMemo(
    () => ({
      productInfo,
      setProductInfo,
    }),
    [productInfo?.product]
  );

  return <ProductContext.Provider value={memoisedValue}>{children}</ProductContext.Provider>;
};

export default ProductContext;
