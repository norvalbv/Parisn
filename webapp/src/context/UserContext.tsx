import { createContext, useState, useMemo, useEffect } from 'react';
import { AuthenticationDetails, CognitoUser, CognitoUserPool } from 'amazon-cognito-identity-js';
import { UserContextInformation, UserInformation } from '../types';

const poolData = {
  UserPoolId: 'eu-west-2_cuz8GIlYg',
  ClientId: '3gh7ll3fdvsq6nh642g06emka6',
};

// const userPool = new CognitoUserPool(poolData);

type ProductContextProviderProps = {
  children?: JSX.Element;
};

const UserContext = createContext<UserContextInformation | null>(null);

export const UserInformationProvider = ({ children }: ProductContextProviderProps) => {
  const [user, setUser] = useState<UserInformation | null>({
    id: '1231287312',
    firstName: 'Benji',
    lastName: 'Norval',
    username: 'BenjiTheGreat',
    email: 'benjinorval@gmail.com',
    image:
      'https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.henley.ac.uk%2Fnews%2F2021%2Fapplied-entrepreneurship-students-seek-support-for-start-ups&psig=AOvVaw1RhLQWkTshWUg_ndVOhtN5&ust=1670189129240000&source=images&cd=vfe&ved=0CA8QjRxqFwoTCKCd3L-x3vsCFQAAAAAdAAAAABAl',
  });

  // const login = async (Username: string, Password: string) => {
  //   return await new Promise((resolve, reject) => {
  //     const user = new CognitoUser({
  //       Username,
  //       Pool: userPool,
  //     });

  //     const authDetails = new AuthenticationDetails({
  //       Username,
  //       Password,
  //     });

  //     user.authenticateUser(authDetails, {
  //       onSuccess: (data) => {
  //         console.log('on succ', data);
  //         resolve(data);
  //       },
  //       onFailure: (err) => {
  //         console.error('on fail', err);
  //         reject(err);
  //       },
  //       newPasswordRequired: (data) => {
  //         console.log('new', data);
  //         resolve(data);
  //       },
  //     });
  //   });
  // };

  useEffect(() => {
    const retreviedProductInfo = localStorage.getItem('userInformation');
    const parsedData: UserInformation = JSON.parse(retreviedProductInfo || 'null');

    setUser(parsedData ?? null);
  }, []);

  useEffect(() => {
    localStorage.setItem('userInformation', JSON.stringify(user));
  }, [user]);

  const memoisedValue = useMemo(
    () => ({
      user: {
        id: user?.id,
        firstName: user?.firstName,
        lastName: user?.lastName,
        username: user?.username,
        email: user?.email,
        image: user?.image,
      },
      setUser,
    }),
    [user]
  );

  return <UserContext.Provider value={memoisedValue}>{children}</UserContext.Provider>;
};

export default UserContext;
