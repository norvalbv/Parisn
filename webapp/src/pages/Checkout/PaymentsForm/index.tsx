import React, { Dispatch, FormEvent, ReactElement, SetStateAction, useState } from 'react';
import { useStripe, useElements, PaymentElement } from '@stripe/react-stripe-js';

type PaymentsFormProps = {
  productId: string;
  setStage: Dispatch<SetStateAction<number>>;
};

const PaymentsForm = ({ productId, setStage }: PaymentsFormProps): ReactElement => {
  const stripe = useStripe();
  const elements = useElements();

  const [formError, setFormError] = useState<string | undefined>();

  const handleSubmit = (event: FormEvent): void => {
    event.preventDefault();

    (async (): Promise<void> => {
      if (!stripe || !elements) return;
      setStage(1);

      const result = await stripe.confirmPayment({
        elements,
        confirmParams: {
          return_url: `${window.location.origin}/payment-successful?product_id=${productId}`,
        },
      });

      if (result.error) {
        setFormError(result.error.message);
      }
    })().catch(() => {});
  };

  return (
    <div className="mockup-window mx-auto w-[50rem] rounded-lg border bg-base-300">
      <div className="flex justify-center bg-base-200 px-4 py-16">
        <form onSubmit={handleSubmit}>
          <PaymentElement />
          <span className="text-xs text-error">{formError}</span>
          <button
            disabled={!stripe}
            className="daisy-btn-primary daisy-btn mt-10 w-40"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  );
};

export default PaymentsForm;
