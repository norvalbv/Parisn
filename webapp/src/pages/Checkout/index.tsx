import { ReactElement, useState } from 'react';
import Button from '../../components/Button';
import Counter from '../../components/Counter';
import useProduct from '../../hooks/useProduct';
import Form from '../../components/Form';
import { checkoutSchema } from '../../utils/validation';
import { useNavigate } from 'react-router-dom';

const Checkout = (): ReactElement => {
  const navigate = useNavigate();
  const { productInfo } = useProduct();

  const truthyDataParsed = productInfo && Object.values(productInfo || '').every((item) => item);

  const [hovered, setHovered] = useState(false);
  console.log(hovered);

  if (!productInfo) return <></>;

  return (
    <>
      {truthyDataParsed ? (
        <>
          <h2 className="text-4xl top-10 border-b w-min mx-auto relative">Checkout</h2>

          <div className="flex justify-around items-center w-4/5 mx-auto h-screen">
            <div className="flex-1 flex flex-col items-center gap-5 pr-5">
              <p className="underline underline-offset-2">Purchase Information:</p>
              <img src={productInfo.product?.Image} className="mx-8 flex-1 rounded-lg" />
              <div className="flex text-sm divide-x">
                <p className="px-2">£{productInfo.price}</p>
                <p className="px-2">{productInfo.product?.Title}</p>
                <p className="px-2">{productInfo.product?.Title}</p>
              </div>
            </div>
            <div className="flex-1 pl-5">
              <Form
                formValues={{
                  firstName: { initialValue: '', type: 'text', label: 'First Name' },
                  lastName: { initialValue: '', type: 'text', label: 'Last Name' },
                  email: { initialValue: '', type: 'email', label: 'Email' },
                  firstLineOfAddress: {
                    initialValue: '',
                    type: 'text',
                    label: 'First Line of Address',
                  },
                  secondLineOfAddress: {
                    initialValue: '',
                    type: 'text',
                    label: 'Second Line of Address',
                  },
                  Country: { initialValue: '', type: 'text', label: 'Country' },
                  Postcode: { initialValue: '', type: 'text', label: 'Postcode' },
                }}
                footerButton={{
                  active: true,
                  label: !hovered ? 'Clear shopping basket' : 'Are you sure?',
                  onClick: () => {
                    if (!hovered) setHovered(true);
                    else {
                      localStorage.clear();
                      setHovered(false);
                      navigate('/');
                    }
                  },
                  onMouseLeave: () => setHovered(false),
                  showPopup: {
                    active: true,
                    text: 'Note: If you clear your basket, you will not be able to add anything new for 60 seconds',
                  },
                }}
                submitButton={{ label: `Purchase for £${productInfo.price}`, className: 'mt-4' }}
                submitFn={(values) => {
                  console.log(values);
                }}
                validationSchema={checkoutSchema}
              />
              <Counter />
              {/* 
                  Note: If you clear your basket, you will not be able to add anything new for 60
                  seconds
        */}
            </div>
          </div>
        </>
      ) : (
        <div className="h-screen w-full flex flex-col gap-4 justify-center items-center text-xl">
          Basket Empty
          <Button text="Continue Shopping" navigateTo="/catalogue" />
        </div>
      )}
    </>
  );
};

export default Checkout;
