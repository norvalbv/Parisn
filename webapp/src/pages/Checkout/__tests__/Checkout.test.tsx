import { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Checkout from '../';
import { ProductContextProvider } from '../../../context/ProductContext';

describe('<Checkout />', () => {
  test('Checkout props with mandatory props', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <ProductContextProvider>
            <Checkout />
          </ProductContextProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
