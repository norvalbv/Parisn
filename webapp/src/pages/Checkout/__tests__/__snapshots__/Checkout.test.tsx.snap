// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkout /> Checkout props with mandatory props 1`] = `
<div>
  Checkout
</div>
`;
